src/modm/driver/can/mcp2515_impl.hpp

```diff
@@ -186,22 +203,73 @@ modm::Mcp2515<SPI, CS, INT>::setMode(Can::Mode mode)
 }
 
 // ----------------------------------------------------------------------------
+
 template <typename SPI, typename CS, typename INT>
 bool
 modm::Mcp2515<SPI, CS, INT>::isMessageAvailable()
 {
-	return !interruptPin.read();
+	return rxQueue.isNotEmpty();
+}
+
+template <typename SPI, typename CS, typename INT>
+bool
+modm::Mcp2515<SPI, CS, INT>::getMessage(can::Message& message, uint8_t* /*filter_id*/)
+{
+	if (rxQueue.isEmpty())
+	{
+		// no message in the receive buffer
+		return false;
+	}
+	else {
+        auto& rxMessage = rxQueue.get();
+		memcpy(&message, &rxMessage, sizeof(message));
+		rxQueue.pop();
+		return true;
+	}
+}
+
+// ----------------------------------------------------------------------------
+template <typename SPI, typename CS, typename INT>
+bool
+modm::Mcp2515<SPI, CS, INT>::isReadyToSend()
+{
+	return txQueue.isNotFull();
 }
 
 // ----------------------------------------------------------------------------
 template <typename SPI, typename CS, typename INT>
 bool
-modm::Mcp2515<SPI, CS, INT>::getMessage(can::Message& message)
+modm::Mcp2515<SPI, CS, INT>::sendMessage(const can::Message& message)
 {
+	if (not modm_assert_continue_ignore(txQueue.push(message), "mcp2515.can.tx",
+			"CAN transmit software buffer overflowed!", 1)) {
+		return false;
+	}
+	return true;
+}
+
+// ----------------------------------------------------------------------------
+template <typename SPI, typename CS, typename INT>
+void
+modm::Mcp2515<SPI, CS, INT>::mcp2515interrupt(){
+	using namespace mcp2515;
+
+	can::Message message;
+	bool success = mcp2515readMessage(message);
```

Comment:
This is not safe to be called from an interrupt context. You could be within `update()` transmitting a message while the EXTI interrupt triggers and have a race condition on the SPI. The only safe way to access the SpiMaster from an interrupt is exclusively use it from that handler.

Reply:
ok so i will change the receiving branch to read via the rx queue as well

Reply:
the rx queue is now also processed from the update function; i removed the interrupt entirely because there is no point now :D